src/Interpolators.f90

```diff
@@ -0,0 +1,162 @@
+  REAL(SRK) :: data_p,f1(2),f2(2)
+  INTEGER(SIK) :: i,j,N_i,N_j,i_p,j_p
+
+  CALL Get_points_and_weights(labels1,axis_p(1),f1,i_p,N_i)
+  CALL Get_points_and_weights(labels2,axis_p(2),f2,j_p,N_j)
+
+  !Get interpolated value
+  data_p=0.0_SRK
+  DO j=MAX(j_p-1,1),MIN(j_p,N_j)
+    DO i=MAX(i_p-1,1),MIN(i_p,N_i)
+      data_p=data_p+f1(i-i_p+2)*f2(j-j_p+2)*table(i,j)
+    ENDDO
+  ENDDO
+
+ENDFUNCTION Interp_2D
+!
+!-------------------------------------------------------------------------------
+!> @brief This routine takes a 2D data table and linearly interpolates to find
+!>        a desired point.
+!> @param labels1 axis labels at which data points in table are defined for 1st dimension
+!> @param labels2 axis labels at which data points in table are defined for 2nd dimension
+!> @param labels3 axis labels at which data points in table are defined for 3rd dimension
+!> @param table contains data to be interpolated between defined at points
+!>        given in labels1, labels2, and labels 3
+!> @param axis_p point at which interpolation needs to be done to return a value
+!> @param data_p interpolated data found through linear interpolation to the point axis_p
+!>
+FUNCTION Interp_3D(labels1,labels2,labels3,table,axis_p) RESULT(data_p)
+  REAL(SRK),INTENT(IN) :: labels1(:),labels2(:),labels3(:),table(:,:,:),axis_p(:)
+  REAL(SRK) :: data_p,f1(2),f2(2),f3(2)
+  INTEGER(SIK) :: i,j,k,N_i,N_j,N_k,i_p,j_p,k_p
+
+  CALL Get_points_and_weights(labels1,axis_p(1),f1,i_p,N_i)
+  CALL Get_points_and_weights(labels2,axis_p(2),f2,j_p,N_j)
+  CALL Get_points_and_weights(labels3,axis_p(3),f3,k_p,N_k)
+
+  !Get interpolated value
+  data_p=0.0_SRK
+  DO k=MAX(k_p-1,1),MIN(k_p,N_k)
+    DO j=MAX(j_p-1,1),MIN(j_p,N_j)
+      DO i=MAX(i_p-1,1),MIN(i_p,N_i)
+        data_p=data_p+f1(i-i_p+2)*f2(j-j_p+2)*f3(k-k_p+2)*table(i,j,k)
+      ENDDO
+    ENDDO
+  ENDDO
+
+ENDFUNCTION Interp_3D
+!
+!-------------------------------------------------------------------------------
+!> @brief This routine takes a 1D array of table labels (the points along a single
+!>        dimension at which table entries are defined) and the interpolation point
+!>        coordinate along that dimension and returns the index of the label entry
+!>        directly ahead of that point along with the weights of that point and the
+!>        preceeding point.
+!> @param labels axis labels at which data points in table are defined
+!> @param axis_p point at which interpolation needs to be done to return a value
+!> @param f weights associated with the points on either side of axis_p
+!> @param i_p index of point directly ahead of axis_p in labels
+!> @param N_i number of entries of labels
+!>
+SUBROUTINE Get_points_and_weights(labels,axis_p,f,i_p,N_i)
+  REAL(SRK),INTENT(IN) :: labels(:),axis_p
+  REAL(SRK),INTENT(OUT) :: f(2)
+  INTEGER(SIK),INTENT(OUT) :: i_p,N_i
+  INTEGER(SIK) :: i
+  N_i=SIZE(labels)
+
+  !Determine labels index(s) and weights
+  f(:)=1.0_SRK
```

Comment:
how can both of these weights be 1?  Should it be 1,0?

Reply:
If both weights need to be set they are. If only one needs to be set, i.e. when there is an out of range "extrapolation" needed then the loop/index magic in the 1D, 2D, and 3D functions handles that. In other words there is no need to find out which weight needs to be 0.